### PR TITLE
8330274: C2 SuperWord: VPointer invar: same sum with different addition order should be equal

### PR DESCRIPTION
**Pull request by eme64**
This is an enhancement for AutoVectorization.

I want to improve the detection of `invar`s that are equivalent (guaranteed to compute the same value), but don't have the identical node (the computation is in a different order).

Note: only about 100 lines are real changes, the rest is tests. These are the first tests that check vectorization for MemorySegments.

**Solution Sketch: "canonicalize" the invar**

- Extract all summands of the `invar`: make a list.
  - Parse through `AddL`, `SubL`, `AddI`, `SubI`, to get summands.
  - Bypass `CastLL` and `CastII`
  - Recursively treat `ConvI2L`, `LShiftI` and `LShiftL`: i.e. canonicalize their input.

- Sort all extracted summands by node idx.
- Add up all summands in new order.

If two `invar`s use the same summands, then we know that after canonicalization the new nodes representing the `invar`s must be the same.

**Example**

```
invar1 = b + c + d + a
invar2 = d + b + a + c
```
-> equivalent but not identical nodes

Sort, and add up again:
```
invar1 = a + b + c + d
invar2 = a + b + c + d
```
-> now the nodes are identical

**Motivation: MemorySegment with invar**

One might think that this is a big of a special case: why would anybody write indices to an Array or MemorySegment where the invar has a different addition order for its summands?

This example did not vectorize, even though it should:
https://github.com/openjdk/jdk/blob/78e42d6e311c33548d16c6c74493388d9850238e/test/hotspot/jtreg/compiler/loopopts/superword/TestEquivalentInvariants.java#L425-L441

Both the `get` and the `set` look like they have the same address, and the address increases by a byte in each iteration.

Upon inspection, I saw that the `invar` that `VPointer` produces for the two operations are not identical: the order of addition of the `invar`'s summands is different, and thus the `invar` nodes are different.

The consequence: Only if we can prove that the two `invar` are identical can we know that the addresses are identical, and that there is no aliasing for loop carried dependencies. Since we have different `invar`, we don't know how the two addresses alias, and that prevents vectorization.

Why does this happen? After parsing, the graph looks like this:
image

We already see that the two addresses are different only by a `CastLL`, with type `long:>=0`. Somehow, that was only deduced for the load, and not the store.
```
load_adr = base + memory_segment_offset + CastLL(invar + iv)
store_adr = base + memory_segment_offset + invar + iv
```

And when we run SuperWord, the graph looks like this:
image

The invar now has 3 summands:
- `530 LoadL`: the offset that the MemorySegment has internally, to indicate its offset to the base of the beginning of the backing array.
- `11 Para`: the input `invar` parameter from our test method.
- `1460 Phi` pre-loop iv increment (value unknown at compile time)

We can see that the addition of the summands is quite different for the load and store.

An alternative solution would have been to add a corresponding `CastLL` for both the `load` and `store`, and hope that this means that the final address would look identical, though I don't know how difficult that would be.

**Future Work**

There are still a few relevant `MemorySegment` address patterns that are not properly recognized by `VPointer`.

In some cases, the `RangeCheck` is not eliminated from the loop: [JDK-8327209](https://bugs.openjdk.org/browse/JDK-8327209)

I also have seen a case where the `If` from the `VarHandleSegmentAsInts::offsetPlain` did not fold away.
image
The check looks like `((2271 AddL) + 4) & 3 == 0`. Maybe this could be fixed with IGVN optimizations, but I'm not sure.

I am also thinking about refactoring `VPointer` completely. It has grown over time, and its pattern-matching is a bit nasty.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330274](https://bugs.openjdk.org/browse/JDK-8330274): C2 SuperWord: VPointer invar: same sum with different addition order should be equal (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18795/head:pull/18795` \
`$ git checkout pull/18795`

Update a local copy of the PR: \
`$ git checkout pull/18795` \
`$ git pull https://git.openjdk.org/jdk.git pull/18795/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18795`

View PR using the GUI difftool: \
`$ git pr show -t 18795`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18795.diff">https://git.openjdk.org/jdk/pull/18795.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18795#issuecomment-2066241198)

### CODE REVIEW

File: src/hotspot/share/opto/vectorization.cpp [hunk 1]
```diff
@@ -98,6 +98,7 @@ VStatus VLoop::check_preconditions_helper() {
       tty->print("  cl_exit->in(0) %d", _cl_exit->in(0)->_idx); _cl_exit->in(0)->dump();
       tty->print("  lpt->_head %d", _cl->_idx); _cl->dump();
       _lpt->dump_head();
+      _cl->dump_bfs(100, _cl_exit, "c-");
```
Comment (eme64) [inline src/hotspot/share/opto/vectorization.cpp:101]: Note: Simply makes it easier to see what kind of CFG is between the loop head and end. Is it a `RangeCheck`, or some `If`?

**Comment by eme64** [comment]
@robcasloz I discussed it a little with @chhagedorn.

We think it would be tricky to get such a IGVN optimization right, and possibly it would not work for all cases.

Basically, we would have some value `v` with a direct use `use1` and an indirect use `use2` via some `CastLL`.
```
    v
 +-+-------+
 |         |
use1      CastLL (with dependency, and constrained type)
           |
          use2
```

The `CastLL` has a `If` dependency, for some `RangeCheck` for example, and therefore it constrains also the type of `v`.

To get `use1` and `use2` to have the same input, we would either have to:
- Remove the `CastLL`: not a good idea, this would lose type info and in some cases the lost pinning would lead some nodes further down to float up.
- Make `use1` use the `CastLL` instead of `v` directly. But then we would need to prove that all uses of `use1` are under the same dependency as the `CastLL`, and I think that would be tricky and error prone. And probably we could not make it work in all cases.

I would really like `SuperWord` to be robust to `CastII/LL`, and this optimization here does that. It also allows to parse more cases, i.e. `a + b + c` and `b + c + a`. I suspect that this could happen to a Java user, that they accidentally swap the addition order of an index (though maybe rare).

@robcasloz What do you think?

**Comment by robcasloz** [comment]
Fair enough, I agree that, even if there was a solution for this specific case, we would probably encounter other cases where GVN would not be powerful enough to detect the equivalence. I still wonder though what could be causing the `CastLL(invar + iv)` vs. `invar + iv` divergence in your example and whether anything could be done to get rid of it. Maybe worth filing a RFE for further investigation.

**Comment by eme64** [comment]
@robcasloz are you intending to review, or was that just a drive-by comment/question?

**Comment by robcasloz** [comment]
> @robcasloz are you intending to review, or was that just a drive-by comment/question?

I'm happy to review this, just give me a few days.

File: src/hotspot/share/opto/vectorization.hpp [hunk 2]
```diff
@@ -723,6 +723,9 @@ class VPointer : public ArenaObj {
   VPointer(VPointer* p);
   NONCOPYABLE(VPointer);
 
+  Node* convI2L(Node* n);
```
Comment (eme64) [inline src/hotspot/share/opto/vectorization.hpp:726]: TODO: remove

**Comment by eme64** [comment]
@robcasloz I decided to drop this PR. Rather than making the current `VPointer` more complicated, I want to refactor it, and then allow for extensions like this to be much more integrated into the pattern matching.

See:
[JDK-8330991](https://bugs.openjdk.org/browse/JDK-8330991) C2 SuperWord: refactor VPointer

**Pull request closed without merging**